components/merkledb/src/db.rs

```diff
@@ -642,26 +726,92 @@ impl<'a> IndexAccess for &'a Fork {
     }
 }
 
-impl IndexAccess for Rc<Fork> {
-    type Changes = ChangesRef<'static>;
+impl RawAccess for Rc<Fork> {
+    type Changes = ChangesMut<'static>;
 
     fn snapshot(&self) -> &dyn Snapshot {
         &self.patch
     }
 
     fn changes(&self, address: &IndexAddress) -> Self::Changes {
         let changes = self.working_patch.take_view_changes(address);
-        ChangesRef {
+        ChangesMut {
             changes,
             key: address.clone(),
             parent: WorkingPatchRef::Owned(Self::clone(self)),
         }
     }
 }
 
-impl AsRef<dyn Snapshot> for Fork {
-    fn as_ref(&self) -> &dyn Snapshot {
-        &self.patch
+/// Readonly wrapper for a `Fork`.
+///
+/// This wrapper allows to read from index state from the fork
+/// in a type-safe manner (it is impossible to accidentally modify data in the index), and
+/// without encountering runtime errors when attempting to concurrently get the same index
+/// more than once.
+///
+/// Since the wrapper borrows the `Fork` immutably, it is still possible to access indexes
+/// in the fork directly. In this scenario, the caller should be careful that `ReadonlyFork`
+/// does not access the same indexes as the original `Fork`: this will result in a runtime
+/// error (sort of like attempting both an exclusive and a shared borrow from a `RefCell`
+/// or `RwLock`).
+///
+/// # Examples
+///
+/// ```
+/// # use exonum_merkledb::{access::AccessExt, Database, ReadonlyFork, TemporaryDB};
+/// let db = TemporaryDB::new();
+/// let fork = db.fork();
+/// fork.as_ref().get_list("list").push(1_u32);
+/// let readonly: ReadonlyFork<'_> = fork.readonly();
+/// let list = readonly.get_list::<_, u32>("list");
+/// assert_eq!(list.get(0), Some(1));
+/// let same_list = readonly.get_list::<_, u32>("list");
+/// // ^-- Does not result in an error!
+///
+/// // Original fork is still accessible.
+/// let mut map = fork.as_ref().get_map("map");
+/// map.put(&1_u32, "foo".to_string());
+/// ```
+///
+/// There are no write methods in indexes instantiated from `ReadonlyFork`:
+///
+/// ```compile_fail
+/// # use exonum_merkledb::{access::AccessExt, Database, ReadonlyFork, TemporaryDB};
+/// let db = TemporaryDB::new();
+/// let fork = db.fork();
+/// let readonly: ReadonlyFork<'_> = fork.readonly();
+/// let mut list = readonly.get_list("list");
+/// list.push(1_u32); // No `push` method for `ListIndex<ReadonlyFork, u32>`!
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ReadonlyFork<'a>(&'a Fork);
+
+impl<'a> ToReadonly for &'a Fork {
+    type Readonly = ReadonlyFork<'a>;
+
+    fn to_readonly(&self) -> Self::Readonly {
```

Comment:
I'm not sure, `ToReadonly` looks like a conversion trait, so `to_*` naming looks natural to me (or maybe `as_readonly`, [since conversion is cheap](https://doc.rust-lang.org/1.0.0/style/style/naming/conversions.html) for supported types).

Reply:
I'd suggest to use `read` name. And rename trait to `Read` or something like `FooRead`

Reply:
There is `Read` trait in std, which has substantially different semantics. IMO, `Read` implies that you can read something from the implementation, which is not the case here; the trait performs a *conversion*, so IMO, naming `ToReadonly::to_readonly` or `AsReadonly::as_readonly` conveys the meaning best.